### ISSUE TITLE
Numerical differences between self.test_session and regular sessions?

### ISSUE THREAD

**Issue by alquraishi** [tensorflow/tensorflow#3574]
I'm trying to use the test suite to create tests that verify various aspects of my model. Strangely, I'm getting numerical differences between the exact same model run on the exact same machine if I use `self.test_session` inside `tf.test.TestCase` versus if I just run the same piece of code in a regular session. Most models work fine, it's just that certain flavors that use dropout show big discrepancies, usually after 4 or 5 steps of training. Is this expected behavior? If so what's the cause of the discrepancy and can it be removed?

**Comment by vrv** [comment]
The code for test_session isn't too long: https://github.com/tensorflow/tensorflow/blob/73ced9d797056c7e67a06ed2098dd809d85ec44a/tensorflow/python/framework/test_util.py#L239 

mostly involves making device placement a little bit easier in tests.  The only thing that comes to mind is that we turn off some of the compiler optimizations in tests, and so perhaps the numerical values are slightly different.  Try commenting out the corresponding line in test_session and let us know if that's the problem!

**Comment by alquraishi** [comment]
Are you referring to this line?

https://github.com/tensorflow/tensorflow/blob/73ced9d797056c7e67a06ed2098dd809d85ec44a/tensorflow/python/framework/test_util.py#L251

Where the optimizer options are set? I did comment it out but it had no effect.

**Comment by vrv** [comment]
Okay, that's good to know.   The only other thing that the function does is deal with device placement, so it would be good to see the difference between your examples by finding a difference using log_device_placement=True.  That is, test_session() and tf.Session() with log_device_placement=True might be different in your use cases, and that might be highlighting either a bug, or numerical differences.

Can you provide a small test case that we can use to validate?

**Comment by yaroslavvb** [comment]
There's also "enable_soft_placement" -- in 0.9,
tf.Session(tf.ConfigProto()) gave session without soft placement,
tf.Session() gave session with soft placement

**Comment by vrv** [comment]
That's true, but code that runs with hard placement successfully should place ops no differently than code that runs with soft placement (soft placement only does something different if the ops are not implemented on the device the user specified).

**Comment by alquraishi** [comment]
I'm running a Mac version and only with CPU support, so device placement can't be the reason, can it?

**Comment by vrv** [comment]
Hmm, then no :). Not really sure where differences can come from then. We
will need a repro

**Comment by alquraishi** [comment]
I figured out what was going on. I was setting `inter_op_parallelism_threads` and `intra_op_parallelism_threads` to specific values and that wasn't being reflected in `test_session`. I see now that it's possible to pass it an explicit `ConfigProto` object and that solves the problem.
